Data Plane AzureML/Azure-AI - Prompt Asset

diondrapeck opened the pull request:
# Data Plane API - Pull Request

Adding a Prompt subclass of AssetVersion for use with prompt engineering tasks

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Schedule a data plane REST API spec review](https://aka.ms/azsdk/onboarding/restapischedule)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

diondrapeck committed:
Add Prompt models to model.tsp

diondrapeck committed:
Add routes for Prompt

diondrapeck committed:
Add categories attribute to Prompt model

diondrapeck committed:
Fix errors and compile

diondrapeck committed:
Add examples

diondrapeck committed:
Update azure-ai-assets.json

diondrapeck committed:
Update list latest example

diondrapeck committed:
Update attributes and re-generate examples

diondrapeck committed:
Remove unintended example for Indexes

diondrapeck committed:
Add names to examples

diondrapeck committed:
Fix formatting errors with prettier

diondrapeck committed:
Delete ListLatest_MinimumSet example

diondrapeck committed:
Run prettier on azure-ai-assets.json

diondrapeck committed:
Fix name

kseager commented:
could you also remove v2.0 from the generic asset url? Here: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/machinelearningservices/AzureAI.Assets/main.tsp

diondrapeck committed:
Add May 1st preview version

diondrapeck committed:
Revert changes to 4/1 preview

diondrapeck committed:
Add 05/01 examples to AzureAIAssets directory and generate examples again and compile

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Run prettier on examples

diondrapeck committed:
Merge branch 'data-plane-prompt-asset' of https://github.com/diondrapeck/azure-rest-api-specs into data-plane-prompt-asset

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Update azure-ai-assets.json for 4/1 to revert changes

diondrapeck committed:
Merge branch 'data-plane-prompt-asset' of https://github.com/diondrapeck/azure-rest-api-specs into data-plane-prompt-asset

diondrapeck committed:
Remove version 2.0 from generic asset url

diondrapeck committed:
Revert 4/1 changes and remove 2.0 version and format

diondrapeck committed:
Revert generic asset version change for april

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

Commit:
version prompt

diondrapeck committed:
Merge pull request #1 from kseager/version_prompt

Add versioning to prompt

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Make templatePath required

diondrapeck committed:
Merge branch 'data-plane-prompt-asset' of https://github.com/diondrapeck/azure-rest-api-specs into data-plane-prompt-asset

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

Commit:
fix paths

diondrapeck committed:
Merge pull request #3 from kseager/readme-updates-diondrapeck

readme for dataplane fix paths

mikekistler commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+        },
+        "stage": {
+          "type": "string",
+          "description": "Update stage to 'Archive' for soft delete. Default is Development, which means the asset is under development.",
```
If the intent is to support soft delete, you should use an established pattern for this. Otherwise please don't call it soft delete.

mikekistler commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+          "type": "string",
+          "description": "Description information of the asset."
+        },
+        "systemData": {
+          "$ref": "#/definitions/SystemData",
+          "description": "Metadata containing createdBy and modifiedBy information.",
+          "readOnly": true
+        },
```
The nesting here seems gratuitous. I think it would be better to simply put all these properties at the top level.

kseager replied:
not following this, the API has a single layering...

mikekistler replied:
See my suggestion below.

mikekistler commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+            "type": "string"
+          }
+        },
+        "properties": {
+          "type": "object",
+          "description": "Asset's properties.",
+          "additionalProperties": {
+            "type": "string"
+          }
+        },
```
Are these needed? What would they be used for? Why isn't the `tags` field just above sufficient?

kseager replied:
Tags are different then properties. Properties are add-only, whereas tags are fully mutable.

mikekistler replied:
How would the user know this? And why make properties add-only? I think these things need to be described.

diondrapeck replied:
I've updated the descriptions to clarify this

mikekistler commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+          }
+        },
+        "storageUri": {
+          "type": "string",
```
```suggestion
          "type": "string",
          "format": "uri",
```

kseager replied:
we've already committed to string for this

diondrapeck replied:
We confirmed this won't break our code generation. Will change.

mikekistler commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+      "type": "object",
+      "description": "Next version definition.",
+      "properties": {
+        "nextVersion": {
+          "type": "integer",
+          "format": "int64",
+          "description": "Next version as defined by the server."
+        },
```
I recommended above that you simply delete this property, and I think that's the best course of action. But if you do keep it, I think you need to make it a string. It's just weird that version is a string everywhere else in the API except here.

kseager replied:
For this case in particular, when we increment the next version we do a +1 on the previous version so for this case we assume version must be an integer vs string

mikekistler replied:
But all the previous versions are strings. So what if the version is "one" or "my-favorite-version"? This is extremely ill-defined and will only create problems down the road. Please remove it.

mikekistler reviewed (commented):
I left a bunch of comments. Many of these are on the existing "Index" type and I apologize that I missed these in the earlier review. Most of these also apply to the new Prompt type -- I just didn't want to try to flag every instance. So please consider all the comments to apply across the whole API even if they are only on specific lines.

Also, I am still uncomfortable with the versioning approach being used here and I think it warrants a deeper review and discussion. I'm not questioning the need to support versioning, but just the specific approach being used, where there are indexes and versions of indexes that are both actually the same thing, but from the URL structure look like different resources.

It isn't necessary to fix all of these in this API version, but I think most _should_ be addressed before the API goes GA, so fixing them as soon as you can is probably best for your users.

kseager commented:
> I left a bunch of comments. Many of these are on the existing "Index" type and I apologize that I missed these in the earlier review. Most of these also apply to the new Prompt type -- I just didn't want to try to flag every instance. So please consider all the comments to apply across the whole API even if they are only on specific lines.
> 
> Also, I am still uncomfortable with the versioning approach being used here and I think it warrants a deeper review and discussion. I'm not questioning the need to support versioning, but just the specific approach being used, where there are indexes and versions of indexes that are both actually the same thing, but from the URL structure look like different resources.
> 
> It isn't necessary to fix all of these in this API version, but I think most _should_ be addressed before the API goes GA, so fixing them as soon as you can is probably best for your users.

We can re-open the discussion before GA. However, this PR is truly meant to be an iteration off of the last PR. It's unfortunate that these reservations were not discussed more thoroughly during the Index API PR, where I had introduced this model and was hoping to address all concerns at that time in order to shelter this PR from those discussions. 

As mentioned before, this modeling for "assets" was already GA'd and published in ARM as an accepted REST API model for versioning resources: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2024-04-01/mfe.json - for example resource ..data/{name}/versions
Although there is always room for improvement, we already have experience working with customers actively on this versioning model and have a proof point of it working for our purposes. The reason we strongly prefer to adhere to this model is for two reasons 
1) We want to interop with our existing GA'd ARM asset set as much as possible 
2) It's an ideal model for creating lineage because we can keep "copies" of each version and reference them uniquely. 

If you still require a deeper review, then can we please address in our next API version? At least going into that API we will be aware of the pre-existing review requirement. 
Thank you for your understanding in this matter.

kseager commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-04-01-preview/azure-ai-assets.json:
```diff
@@ -13,7 +13,7 @@
   ],
   "x-ms-parameterized-host": {
-    "hostTemplate": "{endpoint}/genericasset/v2.0/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MachineLearningServices/workspaces/{workspaceName}",
+    "hostTemplate": "{endpoint}/genericasset/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MachineLearningServices/workspaces/{workspaceName}",
```
this would be considered a bug fix since that route does not exist anymore

mikekistler replied:
If this route ever existed / was available to customers then this is not a bug fix. This is a breaking change. Please do not do this.

kseager commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+      "type": "object",
+      "description": "Index resource Definition",
+      "properties": {
+        "id": {
+          "type": "string",
+          "description": "Fully qualified resource Id: azureml://workspace/{workspaceName}/indexes/{name}/versions/{version} of the index.",
+          "readOnly": true
+        },
```
For the List scenario we need to be able to identify the resource being referenced and so we need either id or name/version for this purpose.

The reason we don't have index and name in the body is because they would then have to be read-only fields (since they are provided as url input parameters) and that ends up causing us problems in one of our curated layers.

Additionally we were using Id before in our ARM API to denote a reference to an asset and found that alignment to work best for our purposes (however since this is not an ARM asset, it will use a different Id notation).

mikekistler replied:
We could let this go in preview but it really should be fixed before GA.

kseager commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+          "readOnly": true
+        },
+        "stage": {
+          "type": "string",
+          "description": "Update stage to 'Archive' for soft delete. Default is Development, which means the asset is under development.",
+          "default": "Development"
```
the reason for the string is the support custom stages and also align with pre-existing asset types

Commit:
refine latest definition

Commit:
get next version definition

Commit:
update descriptions

Commit:
version breaking changes

kseager commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+            "description": "Name of the index.",
+            "required": true,
+            "type": "string",
+            "pattern": "^[a-zA-Z0-9][a-zA-Z0-9-_]{0,254}$"
```
do you have a recommendation on how we'd manage the versioning for this change for Index?

mikekistler replied:
You probably need to talk to the TypeSpec team on this. If it isn't possible to version this then I'd recommend:
a) leave it alone if it has already GA'ed, or 
b) change it now so that it is right for GA.

diondrapeck replied:
I've updated this for 05-01

kseager commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+            "required": false,
+            "type": "string"
+          },
+          {
+            "name": "tags",
+            "in": "query",
+            "description": "Comma-separated list of tag names (and optionally values). Example: tag1,tag2=value2.",
+            "required": false,
+            "type": "string"
+          },
```
Thanks for letting us know. Given the requirement of odata for filter, this it too much work to incorporate into this PR

mikekistler replied:
Okay, but please plan to make this change before GA.

kseager commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+        }
+      },
+      "put": {
+        "operationId": "Indexes_CreateOrUpdate",
+        "description": "Creates or updates a IndexVersion.",
```
do you have a recommendation on how we'd manage the versioning for this change for Index?

mikekistler replied:
I think there are a few options:
- You can deprecate the PUT starting in this version and add PATCH starting in this version.
- You can change this operation to "CreateOrReplace" in this version and deprecate the prior.

diondrapeck committed:
Merge pull request #4 from kseager/review-comments-addressed

Address Spec-Reviewer Comments Prompt review

diondrapeck committed:
Run typespec validation

diondrapeck committed:
Resolve merge conflict

mikekistler reviewed (changes requested):
I really think you need to fix this PR to not break or even change released API versions.

mikekistler commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+          "type": "string",
+          "description": "Description information of the asset."
+        },
+        "systemData": {
+          "$ref": "#/definitions/SystemData",
+          "description": "Metadata containing createdBy and modifiedBy information.",
+          "readOnly": true
+        },
```
```suggestion
        "createdAt": {
          "type": "string",
          "format": "date-time",
          "description": "The timestamp the resource was created at.",
          "readOnly": true
        },
        "createdBy": {
          "type": "string",
          "description": "The identity that created the resource.",
          "readOnly": true
        },
        "createdByType": {
          "type": "string",
          "description": "The identity type that created the resource.",
          "readOnly": true
        },
        "lastModifiedAt": {
          "type": "string",
          "format": "date-time",
          "description": "The timestamp of resource last modification (UTC)",
          "readOnly": true
        }
```

mikekistler reviewed (changes requested):
I left some additional feedback.

Please set up a meeting if any of this is unclear.

diondrapeck committed:
Add uri format to storageUri

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Revert changes to april version

diondrapeck commented:
> I really think you need to fix this PR to not break or even change released API versions.

It didn't at the time I asked you to review - the only new change was removing the 2.0 from the route (which was an approved breaking change in another PR that I simply added here). After addressing some of the comments from the first round, I'm seeing tests failures, so I'll get things back green.

diondrapeck committed:
Add uri format to storageUri and update pattern max length

diondrapeck committed:
Revert unintended changes

diondrapeck committed:
Add 05-01 api version to main.tsp and update add/remove decorators

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Run prettier on azure-ai-assets.json

diondrapeck committed:
Run prettier on azure-ai-assets.json

diondrapeck committed:
Merge branch 'data-plane-prompt-asset' of https://github.com/diondrapeck/azure-rest-api-specs into data-plane-prompt-asset

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Run prettier on examples

diondrapeck committed:
Merge branch 'data-plane-prompt-asset' of https://github.com/diondrapeck/azure-rest-api-specs into data-plane-prompt-asset

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Update dataUri format to be uri

diondrapeck committed:
Update examples

diondrapeck committed:
Merge branch 'data-plane-prompt-asset' of https://github.com/diondrapeck/azure-rest-api-specs into data-plane-prompt-asset

diondrapeck committed:
Update examples to use uri format for dataUri and storageUri

diondrapeck committed:
Replace name value for indexes get next version example

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Replace storageUri in examples

Commit:
updates from removing error response

diondrapeck committed:
Merge pull request #6 from kseager/fix_error_issue

updates from removing error response

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Run validation

diondrapeck committed:
Run validation

diondrapeck committed:
Update pattern to use maxLength

diondrapeck committed:
Update pattern to use maxLength

diondrapeck committed:
Revert "Update pattern to use maxLength"

This reverts commit f740a572779d0024852eaa84a7d6a611c6566f19.

diondrapeck committed:
Run prettier

diondrapeck committed:
Improve tags/properties descriptions

diondrapeck committed:
Add explanation for nextVersion

diondrapeck commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1027 @@
+    "/indexes/{name}:getNextVersion": {
+      "post": {
+        "operationId": "Indexes_GetNextVersion",
+        "description": "Get next Index version.",
```
I've added a description.

This behavior is based on the already defined Data service behavior for nextVersion.

diondrapeck committed:
Update model and routes and regenerate azure-ai-assets.json

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Add updates to 4-1

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Merge branch 'data-plane-prompt-asset' of https://github.com/diondrapeck/azure-rest-api-specs into data-plane-prompt-asset

diondrapeck commented:
@mikekistler regarding then nextVersion logic. I've updated the descriptions to lay it out, but I'll make sure to explain here.

Index uses the Data team's backend for service calls. This is not something our team controls and this pattern has been GA for years. Their nextVersion logic is configured as follows
- Any string is allowed to be set as a version (e.g. your example of "one" or "my-favorite")
- At creation time, Data service screens all the version strings to see if any are integer representations
- If so and if that value is greater than the last seen integer value, nextVersion becomes that value + 1. If there has never been any integer value supplied, the default nextVersion is 1.

In sum, nextVersion = max(lastIntegerVersion + 1, 1)

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

diondrapeck committed:
Merge branch 'data-plane-prompt-asset' of https://github.com/diondrapeck/azure-rest-api-specs into data-plane-prompt-asset

mikekistler reviewed (approved):
I flagged one other issue that should be an easy fix.

Approving for preview.

Commit:
remove required

diondrapeck committed:
Merge pull request #7 from kseager/remove_required

remove required

kseager commented on specification/machinelearningservices/data-plane/Microsoft.MachineLearningServices/preview/2024-05-01-preview/azure-ai-assets.json:
```diff
@@ -0,0 +1,1082 @@
+        "stage": {
+          "type": "string",
+          "description": "Update stage to 'Archive' to archive the asset. Default is Development, which means the asset is under development.",
+          "default": "Development"
```
updated

diondrapeck replied:
Closing this since we've now removed the required specification for `stage`

diondrapeck committed:
Merge branch 'main' into data-plane-prompt-asset

weidongxu-microsoft commented on specification/machinelearningservices/AzureAI.Assets/model.tsp:
```diff
@@ -20,8 +21,8 @@ model AssetVersion {
-  @doc("Update stage to 'Archive' for soft delete. Default is Development, which means the asset is under development.")
-  stage: string = "Development";
+  @doc("Update stage to 'Archive' to archive the asset. Default is Development, which means the asset is under development.")
+  stage?: string = "Development";
```
Should it be `@madeOptional` in 2024-05-01? Or it is a bug in 2024-04-01?

kseager replied:
yes it was a bug

weidongxu-microsoft commented on specification/machinelearningservices/AzureAI.Assets/routes.tsp:
```diff
@@ -120,3 +134,118 @@ interface Indexes {
+
+@added(Versions.`2024-05-01-preview`)
+interface Prompts {
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "Waiting for bug fix: https://github.com/Azure/typespec-azure-pr/issues/3739."
```
Are you copy-paste these operations? That issue in "suppress" is resolved. You should be unblocked.

kseager replied:
thank you for letting me know. Given that this is a rather large rewrite at this juncture, could we incorporate this feedback for the next API? thank you!

kseager replied:
followed-up: acceptable to do a follow-up PR since this will not require a new API version

weidongxu-microsoft commented on specification/machinelearningservices/AzureAI.Assets/routes.tsp:
```diff
@@ -120,3 +134,118 @@ interface Indexes {
+      @path
+      name: string;
+
+      @doc("Version of the prompt")
+      @path
+      version: string;
```
Any doc or API on `{version}` (I seems not finding an API to return all versions)? How does customer know which version to send in path?

kseager replied:
Yes this is a user set version, it's up to them to define the versioning

weidongxu-microsoft commented on specification/machinelearningservices/AzureAI.Assets/routes.tsp:
```diff
@@ -120,3 +134,118 @@ interface Indexes {
+
+      @doc("Ordering of list: Please choose orderby value from ['createdAt', 'lastModifiedAt'].")
+      @query
+      orderby?: string;
```
Is it CSV of multiple order, or a single order?

kseager replied:
single orderby value

weidongxu-microsoft commented on specification/machinelearningservices/AzureAI.Assets/routes.tsp:
```diff
@@ -120,3 +134,118 @@ interface Indexes {
+
+      @doc("Comma-separated list of tag names (and optionally values). Example: tag1,tag2=value2.")
+      @query
+      tags?: string;
```
tsp to define CSV, e.g.
```ts
  @query({
    format: "csv",
  })
  tags?: string[];
```

kseager replied:
so it's a string, not an array of strings

weidongxu-microsoft replied:
`@query` with `format` basically describe in typepec syntax, what you are currently describing in the doc.
